dolib/managers/firewalls.py

```diff
@@ -146,3 +126,127 @@ def remove_rules(
+        res = await self._client.fetch_all(endpoint="firewalls", key="firewalls")
+        return [models.Firewall(**firewall) for firewall in res]
+
+    async def get(self, id: str) -> models.Firewall:
+        res = await self._client.request(
+            endpoint="firewalls/{id}".format(id=id), method="get"
+        )
+        return models.Firewall(**res["firewall"])
+
+    async def create(self, firewall: models.Firewall) -> models.Firewall:
+        res = await self._client.request(
+            endpoint="firewalls",
+            method="post",
+            data=firewall.json(
+                include={
+                    "name",
+                    "inbound_rules",
+                    "outbound_rules",
+                    "droplet_ids",
+                    "tags",
+                }
+            ),
+        )
+        return models.Firewall(**res["firewall"])
+
+    async def update(self, firewall: models.Firewall) -> models.Firewall:
+        res = await self._client.request(
+            endpoint="firewalls/{id}".format(id=firewall.id),
+            method="put",
+            data=firewall.json(
+                include={
+                    "name",
+                    "inbound_rules",
+                    "outbound_rules",
+                    "droplet_ids",
+                    "tags",
+                }
+            ),
+        )
+        return models.Firewall(**res["firewall"])
+
+    async def delete(self, firewall: models.Firewall) -> None:
+        await self._client.request(
+            endpoint="firewalls/{id}".format(id=firewall.id), method="delete"
+        )
+
+    async def add_droplets(self, id: str, droplet_ids: List[int]) -> None:
+        await self._client.request(
+            endpoint="firewalls/{id}/droplets".format(id=id),
+            method="post",
+            json={"droplet_ids": droplet_ids},
+        )
+
+    async def remove_droplets(self, id: str, droplet_ids: List[int]) -> None:
+        await self._client.request(
+            endpoint="firewalls/{id}/droplets".format(id=id),
+            method="delete",
+            json={"droplet_ids": droplet_ids},
+        )
+
+    async def add_tags(self, id: str, tags: List[str]) -> None:
+        await self._client.request(
+            endpoint="firewalls/{id}/tags".format(id=id),
+            method="post",
+            json={"tags": tags},
+        )
+
+    async def remove_tags(self, id: str, tags: List[str]) -> None:
+        await self._client.request(
+            endpoint="firewalls/{id}/tags".format(id=id),
+            method="delete",
+            json={"tags": tags},
+        )
+
+    async def add_rules(
+        self,
+        id: str,
+        inbound_rules: List[models.Firewall.InboundRule] = None,
+        outbound_rules: List[models.Firewall.OutboundRule] = None,
+    ) -> None:
+        assert (
+            inbound_rules is not None or outbound_rules is not None
+        ), "inbound_rules or outbound_rules must be defined"
```

Comment:
I'd use ValueError instead. Generally, assert statements are supposed for tests and can be ignored by optimization flag `-O`, for example:

```python
# t.py
assert False
raise Exception(1)
```

```console
$ python t.py
Traceback (most recent call last):
  File "t.py", line 1, in <module>
    assert False
AssertionError

$ python -O t.py
Traceback (most recent call last):
  File "t.py", line 2, in <module>
    raise Exception(1)
Exception: 1
```

Reply:
Yes, i know this. This assertion for test and lib integrations purposes. In runtime exceptions rised from api errors.